src/common/BaseComponent.ts

```diff
@@ -12,7 +12,7 @@ export class BaseComponent<P, S> extends React.Component<P, S> {
 
   private __async: Async;
   private __events: EventGroup;
-  private __disposables: IDisposable[];
+  private __disposables: (IDisposable | (() => void))[];
```

Comment:
Allowing you to pass anything implementing `() => void` inadvertently allows you to pass any arbitrary function as a dispose callback. It doesn't force you to be explicit that this really *is* a `dispose` callback. We should get people into the habit of being explicit, not letting them be lazy.

Also, the moment you accept a different type of input, you now have to test for it everywhere.

In our current application, we have a system which allows us to compose disposables together into new disposables. Since all disposables are of the same format, this works well. It would get weird if we had to handle a hybrid of `IDisposable` instances and bare `function` instances.